### PR TITLE
More performance optimizations

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Review by graemerocher — APPROVED** [review]

File: http/src/main/java/io/micronaut/http/HttpResponse.java [hunk 1]
```diff
@@ -428,4 +428,25 @@ default Cookies getCookies() {
...
         throw new UnsupportedOperationException("Operation not supported on a " + this.getClass() + " response.");
     }
+
+    /**
+     * Returns a mutable response based on this response.
+     * @return the mutable response
+     * @since 4.0.0
+     */
+    default MutableHttpResponse<?> toMutableResponse() {
+        if (this instanceof MutableHttpResponse<?> mutableHttpResponse) {
```
Comment (yawkat) [inline http/src/main/java/io/micronaut/http/HttpResponse.java:438]: just override it in MutableHttpResponse instead

Comment (yawkat) [reply http/src/main/java/io/micronaut/http/HttpResponse.java:438]: oh i see you did that, remove this check then.

File: router/src/main/java/io/micronaut/web/router/DefaultMethodBasedRouteInfo.java [hunk 2]
```diff
@@ -54,6 +54,7 @@
     private final Optional<Argument<?>> optionalFullBodyArgument;
 
     private RequestArgumentBinder<Object>[] argumentBinders;
+    private boolean needsBody;
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/DefaultMethodBasedRouteInfo.java]: final

Comment (yawkat) [reply router/src/main/java/io/micronaut/web/router/DefaultMethodBasedRouteInfo.java]:
```suggestion
    private final boolean needsBody;
```

File: router/src/main/java/io/micronaut/web/router/DefaultMethodBasedRouteInfo.java [hunk 3]
```diff
@@ -160,4 +171,9 @@ public Optional<Argument<?>> getFullBodyArgument() {
     public String[] getArgumentNames() {
         return argumentNames;
     }
+
+    @Override
+    public boolean needsRequestBody() {
+        return needsBody || super.needsRequestBody();
```
Comment (yawkat) [inline router/src/main/java/io/micronaut/web/router/DefaultMethodBasedRouteInfo.java:177]: i think you could fold super.needsRequestBody into needsBody too, call it in the constructor. probably doesn't matter though.

**Review by yawkat — APPROVED** [review]

**Pull request merged**